pkg/controller/datavolume-controller.go

```diff
@@ -250,57 +94,23 @@ type DataVolumeEvent struct {
 	message   string
 }
 
-// ErrInvalidTermMsg reports that the termination message from the size-detection pod doesn't exists or is not a valid quantity
-var ErrInvalidTermMsg = fmt.Errorf("The termination message from the size-detection pod is not-valid")
+// DataVolumeReconciler interface
+type DataVolumeReconciler interface {
+	prepare(log logr.Logger, dv *cdiv1.DataVolume) error
+	updateStatus(log logr.Logger, dv *cdiv1.DataVolume, pvc **corev1.PersistentVolumeClaim, pvcSpec *v1.PersistentVolumeClaimSpec) (reconcile.Result, error)
+	updateStatusPhase(pvc *corev1.PersistentVolumeClaim, dataVolumeCopy *cdiv1.DataVolume, event *DataVolumeEvent) error
+	cleanup(log logr.Logger, dv *cdiv1.DataVolume) error
+}
```

Comment:
Done

Reply:
Looks like you have refactored to just have a single call to `updateStatus()`.  It is fine if you donnt't want to call that function `syncAndUpdateStatus()`.  But I want to confirm if further refactoring work will include splitting that single call into two separate calls as described above?

Reply:
Isn't the new `sync()` func doing the "work" of the controller as you mentioned? it creates the PVC, PVCSpec etc.

Reply:
To be clear, when I think of `sync()` for import/upload, it's pretty simple, it's just creating the pvc.  But for clone, it is obviously way more complicated depending on clone type and cross namespace or not.

The `sync()` that I see implemented thus far just creates a PVC definition (doesn't actually create the resource).

`updateStatus()` would be conscious of all steps performed in `sync()` and updates the status appropriately.  The only k8s resource it creates/updates is the DataVolume in this case.

So back to my question...what do you think of this split and how does it fit into this pr and the total refactoring plan?

Reply:
Okay, now I see that DataVolumeCloneReconciler has a `Reconcile()`

Reply:
I was originally thinking that it would be cleanest to have a "base" Reconcile that calls sync/updateStatus on the "derived" controller.  What do you think about that?

Reply:
Was thinking the "base" reconciler to be as dumb as possible but have many helpers that can be called by "derived" controllers.

Reply:
> To be clear, when I think of `sync()` for import/upload, it's pretty simple, it's just creating the pvc. But for clone, it is obviously way more complicated depending on clone type and cross namespace or not.
> 
> The `sync()` that I see implemented thus far just creates a PVC definition (doesn't actually create the resource).
> 
> `updateStatus()` would be conscious of all steps performed in `sync()` and updates the status appropriately. The only k8s resource it creates/updates is the DataVolume in this case.
> 
> So back to my question...what do you think of this split and how does it fit into this pr and the total refactoring plan?

We better do it as one of the follow-up PRs, but I don't this PR should wait for this change.

Reply:
> I was originally thinking that it would be cleanest to have a "base" Reconcile that calls sync/updateStatus on the "derived" controller. What do you think about that?

Only `sync()`, which does the common logic is implemented by the "base". The derived controllers implements the specific `updateStatus()`. 

`DataVolumeCloneReconciler` overrides the base `Reconcile()` just in order to pass the `prepare()` and `cleanup()` which you said better be callbacks and not in the interface.